spec/spec_helper.rb

```diff
@@ -20,6 +20,13 @@
   c.ignore_request do |request|
     request.uri =~ /__identify__/
   end
+  # Don't try to playback under the RFC6761 .invalid TLD (should always return NXDOMAIN)
```

Comment:
Are these cases for weird IPs important? Why are they here? This seems like premature optimization / handling cases that will never occur for the user.

Reply:
The idea was to verify that the most common typos are caught.  The `invalid` TLD was specifically chosen for the "validation fails when OIDC server domain is typo'd" tests, because the NXDOMAIN reply can't be simulated using VCR. This is both cleaner (IMO) and less brittle than using mock to override a method hidden a few calls deep within the OIDC connector gem.  It'd be really annoying to track down a test case which spontaneously started failing because someone upstream refactored their gem. :)

Reply:
The spec test cases test the database model, which includes the validators.  So most of these weird URLs are to exercise the validator behavior on the most plausible things someone might stick in.  To provide individual explanations of each one:

This set should cause failures in the http URL format validator, all of which are things we want to catch:
* `ftp://fqdn.is.invalid`
  * wrong protocol - ensure that http URL validator catches a protocol error
* `""`
  * empty - ensure that "present" check works properly
* `http:://x`
  *  too many colons (this happened more than once in testing, and not just by me) - caught by http URL validator match
* `fqdn.is.invalid`
  * no protocol specified, just bare hostname - caught by http URL validator
* `http://user:fqdn.is.invalid/`
  * username:password format URL with no hostname; unlikely, but since this is mostly testing the http URL validator, worthwhile to make sure parsing works as expected.

These should pass the http validator but fail the OIDC validator, thus simultaneously acting as positive tests for http and negative tests for OIDC:
* `http://fqdn.is.invalid`
  * test NXDOMAIN, as above
* `http://user@fqdn.is.invalid/`
  * NXDOMAIN with a username provided
* `https://user:pass@fqdn.is.invalid/`
  * NXDOMAIN with a username and password provided, the third of the three most-likely http URL formats
* `https://192.0.2.0`
  * Address from RFC5737 TEST-NET-1, should trigger timeout because it should not exist anywhere routeable, thus validating handling of a connection timeout
* `http://192.0.2.0:65537`
  * too-high port, which should ultimately trigger a socket error (double-typing a port number is not unlikely)
* `https://1.2.3.256`
  * invalid IP address, which could easily happen due to a typo


I could likely drop the basic auth URL format checks, but they're cheap to do and are things that someone could _technically_ use (though I definitely would not).  I'm basically following the "be liberal in what you accept and strict on what you return" philosophy here.  Since it's in the test cases and not the app and the loops will be there either way to test a couple examples both ways, the relatively trivial extra work to catch the majority of potential mistakes feels justified.